### PR TITLE
Add option --resync-interval to scheduler

### LINKED ISSUE #60860: Scheduler should be able to rebuild its cache and compare it with the existing one
We have seen a few issues indicating that scheduler cache becomes stale and no longer reflects the most recent state of the cluster (#42433, #56057, #43806).
Scheduler cache issues can be very subtle to debug. In order to make it easier, we would like to build a mechanism in scheduler to resync its informers and rebuild its cache periodically and then compare the newly rebuilt cache to its existing cache and log error messages when it sees discrepancies. This feature will be disabled by default. We will probably enable it in our tests and in clusters where customers see stale cache issues

**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

/sig scheduling

### PR DESCRIPTION
**Pull request by anfernee**
**What this PR does / why we need it**:
The option specifies the interval as in seconds between scheduler cache
rebuild and comparison with existing cache. The result will be recorded
in logs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #60860

**Release note**:
```release-note
Add option --resync-interval to scheduler
```

### CODE REVIEW

File: cmd/kube-scheduler/app/server.go
Comment (resouer) [inline cmd/kube-scheduler/app/server.go]: Is this prob only for debug purpose? In that case we need to clarify it.

Comment (anfernee) [reply cmd/kube-scheduler/app/server.go]: done.

**Comment by anfernee** [comment]
the test failure is valid. gonna fix it later.

**Comment by anfernee** [comment]
cc @resouer @bsalamat

File: pkg/scheduler/factory/factory.go
Comment (bsalamat) [inline pkg/scheduler/factory/factory.go]: I think it would be useful to keep the existing logic where we can send a signal to scheduler to trigger a cache compare even when `--resync-interval` is not provided.

File: pkg/scheduler/factory/factory.go
Comment (bsalamat) [inline pkg/scheduler/factory/factory.go]: Doesn't the ticker stop shortly after this point when the function returns? Should this be in the `go func()`?

File: pkg/scheduler/factory/factory.go
Comment (resouer) [reply pkg/scheduler/factory/factory.go]: Yes, I also like the signal trigger

File: pkg/scheduler/factory/factory.go
Comment (anfernee) [reply pkg/scheduler/factory/factory.go]: good catch!

File: test/integration/scheduler/volume_binding_test.go
Comment (bsalamat) [inline test/integration/scheduler/volume_binding_test.go]: By doing this, you are creating another scheduler in the test cluster. `initTestScheduler` at line 348 creates the first one. I guess this has happened by a rebase after a recent change in our integration test library.
I think you should change `createConfiguratorWithPodInformer` in integration/util.go and leave the test as it was.

File: cmd/kube-scheduler/app/server.go
Comment (bsalamat) [inline cmd/kube-scheduler/app/server.go]: Given that this is a user visible option, I would clearly mark it as a debugging option. In order to do so, you may want to change the name to "CacheDebugResyncInterval" and the cli arg to "cache-debug-resync-interval".

File: test/integration/scheduler/volume_binding_test.go
Comment (resouer) [reply test/integration/scheduler/volume_binding_test.go]: Yes, please make this change here: https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/util.go#L66-L88

File: pkg/scheduler/factory/factory.go
Comment (anfernee) [reply pkg/scheduler/factory/factory.go]: ok. Will do.

File: pkg/scheduler/factory/factory.go
Comment (bsalamat) [inline pkg/scheduler/factory/factory.go]: I think we can avoid creating the `ticker` when duration is 0.

**Review by bsalamat — COMMENTED** [review]
Thanks, @anfernee! Looks good. Just a minor comment.

**Comment by bsalamat** [comment]
/lgtm

Thanks a lot, @anfernee!

**Comment by resouer** [comment]
Anfernee, I think these two gce tests are flaky, it happens quite a lot today.

**Comment by anfernee** [comment]
@resouer give it another try

**Comment by bsalamat** [comment]
The failing test is not "Required" and has a known issue. It does not need to pass before we can merge this PR.

**Comment by bsalamat** [comment]
@mtaufen Could you please take a look at the component config changes and approve if they look fine?

File: cmd/kube-scheduler/app/server.go
Comment (mtaufen) [inline cmd/kube-scheduler/app/server.go]: nit: try `The interval in seconds at which the local cache is rebuilt and compared against the existing cache. This should only be used for debugging.`

You may also wish to clarify *why* it is only for debugging.

File: cmd/kube-scheduler/app/server.go
Comment (mtaufen) [inline cmd/kube-scheduler/app/server.go]: I think `CacheResyncInterval` may be a better name. Or perhaps `DebugCacheResyncInterval`. Having `Debug` in the middle of the name just feels strange to me.

In general, I think we should avoid adding debug-only options to components. It can easily lead to API bloat. What is the workflow for using this, what is it designed to debug, and what were the alternatives to extending your API with this parameter?

File: pkg/apis/componentconfig/types.go
Comment (mtaufen) [inline pkg/apis/componentconfig/types.go]: Clarify comment to note this is for debugging only.

File: pkg/apis/componentconfig/types.go
Comment (mtaufen) [inline pkg/apis/componentconfig/types.go]: This is just a general comment, not necessary for this PR, but you should move the scheduler config to its own API group. See https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c.

File: pkg/apis/componentconfig/types.go
Comment (mtaufen) [inline pkg/apis/componentconfig/types.go]: Why not make this a metav1.Duration?

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (mtaufen) [inline pkg/apis/componentconfig/v1alpha1/types.go]: same comments regarding debug clarification, and using metav1.Duration

File: pkg/scheduler/factory/factory.go
Comment (mtaufen) [inline pkg/scheduler/factory/factory.go]: you wouldn't have to do multiplication here if you used metav1.Duration

File: pkg/apis/componentconfig/types.go
Comment (mtaufen) [reply pkg/apis/componentconfig/types.go]: (you can register duration flags with FlagSet.DurationVar, and target the internal time.Duration in metav1.Duration)

**Review by mtaufen — CHANGES_REQUESTED** [review]

File: cmd/kube-scheduler/app/server.go
Comment (mtaufen) [reply cmd/kube-scheduler/app/server.go]: If you use metav1.Duration, you can delete "in seconds"

File: cmd/kube-scheduler/app/server.go
Comment (bsalamat) [reply cmd/kube-scheduler/app/server.go]: I agree that we shouldn't add debug-only option, but alternatives are not great. One option is to add this to scheduler policy config. Another one is to hard-code the value and not make it configurable, which could be fine given that we have another mechanism to invoke the process.

**Comment by mtaufen** [comment]
Could you expose a debugging server, make the port configurable, and
disable by default (port 0). Then you can add arbitrary debug endpoints
without significantly growing your production API.

File: pkg/apis/componentconfig/types.go
Comment (anfernee) [reply pkg/apis/componentconfig/types.go]: I am not aware of this option. I will take a look.

File: cmd/kube-scheduler/app/server.go
Comment (anfernee) [reply cmd/kube-scheduler/app/server.go]: How about adding a simple bash script side-by-side to the scheduler container, which simply does `while true; do sleep 120; kill -12 $SCHED_PID; done`. Then we don't need to have this option exposed and keep the simplicity. 

@mtaufen some background, this debug behavior can be triggered by signal. This option simply adds periodical trigger.

**Comment by anfernee** [comment]
@mtaufen @bsalamat  I am leaning towards closing the PR. As I said, you can achieve periodic trigger thru simple script. Let me know your thoughts.

**Comment by bsalamat** [comment]
> I am leaning towards closing the PR. As I said, you can achieve periodic trigger thru simple script. Let me know your thoughts.

I agree with you and Mike. I also think it is relatively easy to trigger periodic checks.
Thanks for your time, @anfernee!

**Pull request closed without merging**
